### DISCUSSION TITLE
Double precision casts are not supported in table create query

### DISCUSSION THREAD

**Discussion by fujiwaraizuho** [sqldef/sqldef#615] [Feature requests]
db: PostgreSQL 14.5
psqldef: v0.13.22

## create table query
```sql
CREATE TABLE test (
  "double" double precision NOT NULL DEFAULT '100'::double precision
);
```

## --export output
found syntax error when parsing DDL "CREATE TABLE "public"."test" (
    "double" double precision NOT NULL DEFAULT '100'::double precision
)": syntax error at position XXX near 'precision'

## Input SQL
none

## Current output
none

## Expected output
no errors.

**Comment by k0kubun** [comment]
Would you be interested in filing a patch to fix `parser/parser.y`? Then `parser/parser.go` can be updated by `make -C parser`.

**Comment by k0kubun** [comment]
Ohh, actually, this is psqldef. We could (and should) fix it differently. So, we use PostgreSQL's original parser, and `database/postgres/parser.go` converts PostgreSQL's syntax tree to sqldef's SQL syntax tree (but it falls back to the old sqldef parser when it sees an unexpected node). Would you like to look into changing `database/postgres/parser.go` to support this feature?

**Comment by fujiwaraizuho** [comment]
Thanks for the reply.
I will try to fix it.

But I'm not very familiar with SQL syntax trees.
Please let me know if you have any tips to help me solve this issue.

**Comment by k0kubun** [comment]
What I usually do is:

1. Prepare a subset of the original SQL so that the old parser of sqldef can parse it, and dump the syntax tree.
2. Dump the syntax tree that the PostgreSQL parser gives with the original SQL.
3. Write code to convert 2 to 1. Then improve the implementation to support the feature missing in 1.

To dump arbitrary Go objects, I always use my very own https://github.com/k0kubun/pp.

**Comment by fujiwaraizuho** [comment]
I am taking some time to understand.

Does sqldef currently support TYPECAST in CREATE TABLE query?
I can't seem to cast `::text` as well as `::double precision`.

**Comment by fujiwaraizuho** [comment]
Also, I edited parser/parser.y and now I can parse `::double precision`, but I don't know how to do it after this!

https://gist.github.com/tomocrafter/e8c3efdfaf48ca7d4659dffe503651d5#file-parser-y-L1446-L1449

**Comment by k0kubun** [comment]
> Does sqldef currently support TYPECAST in CREATE TABLE query?

The output in the issue description suggests that it doesn't :) You could try removing random places to see what feature is actually missing.

> Also, I edited parser/parser.y and now I can parse ::double precision, but I don't know how to do it after this!

It's cool that you found what's missing in parser.y. If it successfully parses the input, can you make it a pull request?

Also, while it's fine to fix it that way too, this is different from what I talked about in https://github.com/k0kubun/sqldef/issues/309#issuecomment-1319589762. psqldef already has the original PostgreSQL parser that can already parse it, which is just not used when the conversion to sqldef's tree is not supported. So, you can just convert the parse result given by the PostgreSQL parser to sqldef's one.

This is an example commit https://github.com/k0kubun/sqldef/commit/4f9c6198c9797f945bb225fcdbebd56ae3b80cee. If you approach the problem correctly, your patch would look like that too. You don't need to update parser.y for the reason explained above.

**Comment by gfx** [comment]
The latest sqldef support it. Closing.
